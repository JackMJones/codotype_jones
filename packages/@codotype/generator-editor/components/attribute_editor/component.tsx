import { AttributeEditorState, attributeEditorReducer, getInitialState, defaultAttribute } from "./reducer";
import * as React from "react";
import { Dispatch } from "react";
import { AttributeEditorAction, AttributeEditorActionType } from "./actions";
import { AttributeList } from "./AttributeList";
import { AttributeForm } from "./AttributeForm";
import { Datatype, Attribute } from "@codotype/types";
import { AttributeListEmptyState } from "./AttributeListEmptyState";

interface AttributeEditorProps {
  attributes: Attribute[];
  supportedDatatypes: Datatype[];
  onChange: (updatedAttributes: Attribute[]) => void;
}

export function AttributeEditor(props: AttributeEditorProps) {
  const [state, dispatch]: [AttributeEditorState, Dispatch<AttributeEditorAction>] = React.useReducer(attributeEditorReducer, getInitialState(props.attributes));

  React.useEffect(() => {
    if (state.lastUpdatedTime !== null) {
      props.onChange(state.attributes)
    }
  }, [state.lastUpdatedTime]);

  return (
    <div className="row">
      <div className="col-lg-12">

        <button
          onClick={() => {
            dispatch({
              type: AttributeEditorActionType.EDIT,
              attribute: defaultAttribute
            })
          }}
        >
          New +
        </button>

        {state.attributes.length === 0 && (
          <AttributeListEmptyState />
        )}

        { state.attributes.length > 0 && (
          <AttributeList
            attributes={state.attributes}
            onEditButtonClick={(attributeToBeEdited: Attribute) => {
              dispatch({
                type: AttributeEditorActionType.EDIT,
                attribute: attributeToBeEdited
              })
            }}
            onRemoveButtonClick={(attributeToBeRemoved: Attribute) => {
              dispatch({
                type: AttributeEditorActionType.REMOVE,
                attribute: attributeToBeRemoved
              })
            }}
          />
        )}

        {state.editorModel !== null && (
          <AttributeForm
            attributes={state.attributes}
            editorModel={state.editorModel}
            supportedDatatypes={props.supportedDatatypes}
            onSubmit={(updatedAttribute: Attribute) => {
              // Creates a new attribute
              if (updatedAttribute.id === null) {
                dispatch({
                  type: AttributeEditorActionType.CREATE,
                  attribute: updatedAttribute
                })
                return;
              }

              // Updates an existing attribute
              dispatch({
                type: AttributeEditorActionType.UPDATE,
                attribute: updatedAttribute
              })
            }}
            onCancel={() => {
              dispatch({
                type: AttributeEditorActionType.CLEAR_EDITOR,
              })
            }}
          />
        )}
      </div>
    </div>
  );
}
