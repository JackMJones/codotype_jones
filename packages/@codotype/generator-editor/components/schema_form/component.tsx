import * as React from "react";
import { Schema } from "@codotype/types";
import { inflateMeta, sanitizeLabel } from "@codotype/util";

// // // //

/**
 * MetaPreviewProps
 */
interface MetaPreviewProps {
  label: string
}

/**
 * MetaPreview
 */
export function MetaPreview(props: MetaPreviewProps) {

  const schemaMeta = inflateMeta(props.label);

  return (
    <tbody>
      <tr>
        <td className="infoCol">
          <i className="fa fa-fw text-secondary fa-question-circle" title='"Label" is the user-facing token for this model - title-casing and spaces.' />
        </td>
        <td>
          Label
        </td>
        <td className='text-success'>{schemaMeta.label || '...'}</td>
      </tr>
      <tr>
        <td></td>
        <td>Label Plural</td>
        <td className='text-success'>{schemaMeta.label_plural || '...'}</td>
      </tr>
      <tr>
        <td className="infoCol">
          <i className="fa fa-fw text-secondary fa-question-circle" title='"Identifier" is the lowecase, underscored token for this model' />
        </td>
        <td>
          Identifier
        </td>
        <td className='text-success'>{schemaMeta.identifier || '...'}</td>
      </tr>
      <tr>
        <td></td>
        <td>Identifier Plural</td>
        <td className='text-success'>{schemaMeta.identifier_plural || '...'}</td>
      </tr>
      <tr>
        <td className="infoCol">
          <i className="fa fa-fw text-secondary fa-question-circle" title='"Camel Case" is the mixed-case whitespace-free token for this model' />
        </td>
        <td>
          Camel Case
        </td>
        <td className='text-success'>{schemaMeta.camel_case || '...'}</td>
      </tr>
      <tr>
        <td></td>
        <td>Camel Case Plural</td>
        <td className='text-success'>{schemaMeta.camel_case_plural || '...'}</td>
      </tr>
      <tr>
        <td className="infoCol">
          <i className="fa fa-fw text-secondary fa-question-circle" title='"Class Name" is title-cased whitespace-free token for this model' />
        </td>
        <td>
          Class Name
        </td>
        <td className='text-success'>{schemaMeta.class_name || '...'}</td>
      </tr>
      <tr>
        <td></td>
        <td>Class Name Plural</td>
        <td className='text-success'>{schemaMeta.class_name_plural || '...'}</td>
      </tr>
    </tbody>
  )
}

// // // //

interface SchemaFormProps {
  schema: Schema;
  onSubmit: (updatedSchema: Schema) => void;
  onCancel: () => void;
}

/**
 * SchemaForm
 * @param props
 */
export function SchemaForm(props: SchemaFormProps) {
  const [label, setLabel] = React.useState<string>(props.schema.label);

  return (
    <div className="row">

      <div className="col-sm-12 col-lg-6 border-right">
        <div className="row">

          <div className="col-sm-12">

            <h4>New Schema</h4>

            <p className="small mt-2 mb-3 text-muted">
              The <strong>Schema</strong> is a building block that describes <i>data</i>.
            </p>

            <p className="small mt-2 mb-3 text-muted">Its <span className='text-success'>Label</span> should be a <strong>singular noun</strong> - whitespace is allowed. The input field will enforce proper capitalization and spacing.</p>

            <small className="mb-2 text-muted">
              <i className="far fa-lightbulb" />
              Try something simple like <code>Movie</code>, or <code>Movie Rating</code>
            </small>

            <input
              className="form-control form-control-lg"
              placeholder="Label"
              value={label}
              onChange={(e) => {
                setLabel(sanitizeLabel(e.currentTarget.value))
              }}
            />

            <p className="small mt-4 text-muted">
              <strong>Codotype</strong> derives additional <span className='text-success'>Tokens</span> to use for things like naming <strong>files</strong>, <strong>folders</strong>, <strong>variables</strong>, and <strong>database tables</strong>. You can rename or remove a schema whenever you like <i className="far fa-laugh" />
            </p>

          </div>
        </div>
      </div>

      <div className="col-sm-12 col-lg-6 d-flex justify-content-center align-items-center flex-column">
        <p className="mb-0 text-muted">Verify these <span className='text-success'>Tokens </span> before proceeding</p>

        <table className="table table-sm mb-0 mt-2">
          <MetaPreview label={label} />
        </table>

      </div>

      <button
        // disabled={!canSubmit(label)}
        onClick={() => {
          props.onSubmit({
            ...props.schema,
            ...inflateMeta(label.trim()),
          });
        }}
      >
        Save
          </button>

      <button onClick={props.onCancel}>Cancel</button>
    </div>
  )
}

// // // //

// mounted () {
//   setTimeout(() => { this.$refs.input_el.focus() }, 500)
// },
// methods: {
//   onKeyEnter() {
// if (this.onKeypressEnter && this.enableSubmit) this.onKeypressEnter()
// }
// </script>

// // // //

// <style type="text/css" scoped>
// .table-sm {
//   font - size: 80%;
// }

// p.small {
//   font - size: 85%;
// }

// .text-purple {
//   color: purple !important;
// }

// td.infoCol {
//   width: 1rem;
// }
// </style>

