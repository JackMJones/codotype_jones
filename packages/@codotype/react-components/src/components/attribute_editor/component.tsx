import * as React from "react";
import { SortableListHeader } from "../sortable_list_header";
import { Attribute, Datatype, DEFAULT_ATTRIBUTE } from "@codotype/types";
import { Droppable, DragDropContext } from "react-beautiful-dnd";
import { AttributeFormModal, AttributeInput } from "./AttributeFormModal";
import { AttributeDeleteModal } from "./AttributeDeleteModal";
import { AttributeListItem } from "./AttributeListItem";
import { AttributeForm } from "./AttributeForm";
import { AttributeListEmpty } from "./AttributeListEmpty";
import uniqueId from "lodash.uniqueid";

// // // //

// TODO - abstract this into a separate module
export function reorder<T>(
    list: T[],
    startIndex: number,
    endIndex: number,
): T[] {
    const result = Array.from(list);
    const [removed] = result.splice(startIndex, 1);
    result.splice(endIndex, 0, removed);
    return result;
}

// // // //

/**
 * disableSubmit
 * @param label
 */
export function disableSubmit(attributeInput: AttributeInput): boolean {
    return (
        attributeInput.label === "" ||
        attributeInput.identifier === "" ||
        attributeInput.datatype === null
    );
}

// // // //

interface AttributeEditorState {
    attributes: Attribute[];
    lastUpdatedAt: null | number;
}

interface AttributeEditorProps {
    attributes: Attribute[];
    supportedDatatypes: Datatype[];
    onChange: (updatedAttributes: Attribute[]) => void;
}

export function AttributeEditor(props: AttributeEditorProps) {
    const [state, setState] = React.useState<AttributeEditorState>({
        attributes: props.attributes,
        lastUpdatedAt: null,
    });
    const [
        attributeInput,
        setAttributeInput,
    ] = React.useState<AttributeInput | null>(null);
    const [
        showingDeleteModal,
        showDeleteModal,
    ] = React.useState<Attribute | null>(null);

    // Sets props.attributes when props.attributes changes
    React.useEffect(() => {
        setState({ ...state, attributes: props.attributes });
    }, [props.attributes]);

    // Fires off props.onChange
    React.useEffect(() => {
        props.onChange(state.attributes);
    }, [state.lastUpdatedAt]);

    // // // //

    return (
        <div className="card">
            <SortableListHeader
                label="Attributes"
                onClick={() => {
                    setAttributeInput({
                        id: DEFAULT_ATTRIBUTE.id,
                        label: DEFAULT_ATTRIBUTE.label,
                        identifier: DEFAULT_ATTRIBUTE.identifier,
                        required: DEFAULT_ATTRIBUTE.required,
                        unique: DEFAULT_ATTRIBUTE.unique,
                        description: DEFAULT_ATTRIBUTE.description,
                        datatype: DEFAULT_ATTRIBUTE.datatype,
                        default_value: DEFAULT_ATTRIBUTE.datatype,
                        datatypeOptions: DEFAULT_ATTRIBUTE.datatypeOptions,
                        locked: DEFAULT_ATTRIBUTE.locked,
                    });
                }}
            />

            {/* Renders AttributeFormModal */}
            {attributeInput !== null && (
                <AttributeFormModal
                    attributeInput={attributeInput}
                    show={attributeInput !== null}
                    disableSubmit={disableSubmit(attributeInput)}
                    // disableSubmit={false}
                    onCancel={() => {
                        setAttributeInput(null);
                    }}
                    onSubmit={() => {
                        setAttributeInput(null);
                        // Insert new Attribute
                        if (attributeInput.id === "") {
                            const newAttribute: Attribute = {
                                ...attributeInput,
                                id: uniqueId("ATTR_"),
                            };
                            setState({
                                lastUpdatedAt: Date.now(),
                                attributes: [...props.attributes, newAttribute],
                            });
                            setAttributeInput(null);
                            return;
                        }

                        // Update existing attribute
                        setState({
                            lastUpdatedAt: Date.now(),
                            attributes: props.attributes.map((a: Attribute) => {
                                if (a.id === attributeInput.id) {
                                    return attributeInput;
                                }
                                return a;
                            }),
                        });
                        setAttributeInput(null);
                    }}
                >
                    <AttributeForm
                        attributes={props.attributes}
                        attributeInput={attributeInput}
                        onChange={(updatedAttributeInput: AttributeInput) => {
                            setAttributeInput(updatedAttributeInput);
                        }}
                        supportedDatatypes={props.supportedDatatypes}
                    />
                </AttributeFormModal>
            )}

            {props.attributes.length > 0 && (
                <React.Fragment>
                    <AttributeDeleteModal
                        show={showingDeleteModal !== null}
                        onClose={() => showDeleteModal(null)}
                        onConfirm={() => {
                            if (showingDeleteModal !== null) {
                                setState({
                                    attributes: props.attributes.filter(
                                        (a: Attribute) => {
                                            return (
                                                a.id !== showingDeleteModal.id
                                            );
                                        },
                                    ),
                                    lastUpdatedAt: Date.now(),
                                });
                            }

                            showDeleteModal(null);
                        }}
                    />

                    <DragDropContext
                        onDragEnd={result => {
                            if (!result.destination) {
                                return;
                            }

                            if (
                                result.destination.index === result.source.index
                            ) {
                                return;
                            }

                            const updatedAttributes = reorder<Attribute>(
                                props.attributes,
                                result.source.index,
                                result.destination.index,
                            );

                            // Invoke props.onChange directly, prevents multiple re-renders
                            props.onChange(updatedAttributes);
                        }}
                    >
                        <Droppable droppableId="attribute-list">
                            {(provided: any) => {
                                return (
                                    <ul
                                        className="list-group list-group-flush"
                                        ref={provided.innerRef}
                                        {...provided.droppableProps}
                                    >
                                        {props.attributes.map(
                                            (a: Attribute, index: number) => {
                                                return (
                                                    <AttributeListItem
                                                        key={a.id}
                                                        attribute={a}
                                                        index={index}
                                                        onClickEdit={(
                                                            attributeToBeEdited: Attribute,
                                                        ) => {
                                                            setAttributeInput({
                                                                ...attributeToBeEdited,
                                                            });
                                                        }}
                                                        onClickDelete={(
                                                            attributeToDelete: Attribute,
                                                        ) => {
                                                            showDeleteModal(
                                                                attributeToDelete,
                                                            );
                                                        }}
                                                    />
                                                );
                                            },
                                        )}
                                        {provided.placeholder}
                                    </ul>
                                );
                            }}
                        </Droppable>
                    </DragDropContext>
                </React.Fragment>
            )}

            {/* Render empty state */}
            {props.attributes.length === 0 && (
                <AttributeListEmpty
                    onClick={() => {
                        setAttributeInput({ ...DEFAULT_ATTRIBUTE });
                    }}
                />
            )}
        </div>
    );
}
