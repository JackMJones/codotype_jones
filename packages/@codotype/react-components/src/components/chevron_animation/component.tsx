import * as React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronRight } from "@fortawesome/free-solid-svg-icons";
import "./styles.scss";

// // // //

interface ChevronAnimationProps {
    active: boolean;
}

/**
 * ChevronAnimation
 * Renders the card-header for AttributeEditor and RelationEditor
 * @param props - see `ChevronAnimationProps`
 */
export function ChevronAnimation(props: ChevronAnimationProps) {
    return (
        <span className="ml-2">
            {props.active && (
                <FontAwesomeIcon
                    icon={faChevronRight}
                    className="animate-shift"
                />
            )}
            {!props.active && <FontAwesomeIcon icon={faChevronRight} />}
        </span>
    );
}
