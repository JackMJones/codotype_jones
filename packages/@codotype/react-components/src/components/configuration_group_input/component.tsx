import React, { FunctionComponent } from "react";
import "./styles.scss";
import {
    OptionType,
    OptionValue,
    OptionValueInstance,
    ConfigurationGroupProperty,
    ConfigurationGroup,
} from "../types";
import { ConfigurationInputChild } from "./ConfigurationInputChild";
import { ConfigurationGroupVariant } from "./ConfigurationGroupVariant";
import { ConfigurationGroupHeader } from "./ConfigurationGroupHeader";
import { ConfigurationInputFormGroup } from "./ConfigurationInputFormGroup";
import { ConfigurationInstanceInput } from "./ConfigurationInstanceInput";
import { MoreInfoLink } from "../more_info_link";
import { ConfigurationCollectionInput } from "./ConfigurationCollectionInput";

// // // //

/**
 * buildConfigurationGroupPropertyValue
 * Accepts a single ConfigurationGroupProperty and recursively produces its associated OptionValue instance
 * @param property - the single ConfigurationGroupProperty for which the OptionValue is being produced
 */
export function buildConfigurationGroupPropertyValue(
    property: ConfigurationGroupProperty,
): OptionValue {
    if (property.type === OptionType.STRING) {
        return property.defaultValue || "";
    }
    if (property.type === OptionType.BOOLEAN) {
        return property.defaultValue || false;
    }
    if (property.type === OptionType.DROPDOWN) {
        return property.defaultValue || "";
    }
    if (property.type === OptionType.COLLECTION) {
        return [buildConfigurationGroupValue(property.properties)];
        // Should return this:
        // return [];
    }
    if (property.type === OptionType.INSTANCE) {
        return buildConfigurationGroupValue(property.properties);
    }
    return "";
}

/**
 * buildConfigurationGroupValue
 * Builds the top-level OptionValueInstance for a single ConfigurationGroup
 * @param properties - array of ConfigurationGroupProperty
 */
export function buildConfigurationGroupValue(
    properties: ConfigurationGroupProperty[],
): OptionValueInstance {
    const initialValue: OptionValueInstance = {};

    // Defines empty ConfigurationGroupValue
    // Iterates over each property in the group and assigns values
    const configurationGroupValue: any = properties.reduce(
        (val, property: ConfigurationGroupProperty) => {
            // Updates val with data for ConfigurationGroupProperty
            if (property.allowDisable && !property.required) {
                val[property.identifier] = {
                    enabled: property.enabled,
                    value: buildConfigurationGroupPropertyValue(property),
                };
            } else {
                val[property.identifier] = buildConfigurationGroupPropertyValue(
                    property,
                );
            }
            // Returns val
            return val;
        },
        initialValue, // Passes in empty OptionValueInstance
    );

    // Returns ConfigurationGroupValue
    return configurationGroupValue;
}

// // // //

interface ConfigurationInputProps {
    value: OptionValueInstance;
    configurationGroup: ConfigurationGroup;
    onChange: (updatedVal: OptionValueInstance) => void;
}
export const ConfigurationInput: FunctionComponent<ConfigurationInputProps> = (
    props: ConfigurationInputProps,
) => {
    return (
        <div className="row mt-3">
            <div className="col-lg-12">
                {/* ConfigurationGroupHeader */}
                <ConfigurationGroupHeader
                    configurationGroup={props.configurationGroup}
                />

                {/* Renders ConfigurationGroup variants */}
                {/* QUESTION - maybe rename to `layout`? */}
                <ConfigurationGroupVariant
                    configurationGroup={props.configurationGroup}
                >
                    {props.configurationGroup.properties.map(
                        (property: ConfigurationGroupProperty) => {
                            if (property.type === OptionType.COLLECTION) {
                                return (
                                    <ConfigurationCollectionInput
                                        label={property.label}
                                        properties={property.properties}
                                        onChange={(updatedVal: OptionValue) => {
                                            props.onChange({
                                                ...props.value,
                                                // [property.identifier]: updatedVal,
                                            });
                                        }}
                                        value={props.value}
                                    />
                                );
                            }

                            // Handle instance input
                            if (property.type === OptionType.INSTANCE) {
                                const val =
                                    // @ts-ignore
                                    props.value[property.identifier];
                                return (
                                    <div
                                        className="card card-body mb-4 mt-2 mx-2"
                                        key={property.identifier}
                                    >
                                        <ConfigurationInstanceInput
                                            label={property.label}
                                            properties={property.properties}
                                            onChange={(
                                                updatedVal: OptionValueInstance,
                                            ) => {
                                                if (property.allowDisable) {
                                                    props.onChange({
                                                        ...props.value,
                                                        [property.identifier]: {
                                                            enabled: true,
                                                            value: updatedVal,
                                                        },
                                                    });
                                                } else {
                                                    props.onChange({
                                                        ...props.value,
                                                        [property.identifier]: updatedVal,
                                                    });
                                                }
                                            }}
                                            value={val}
                                        />
                                        {/* <p>{property.identifier}</p> */}
                                        {/* <p>{JSON.stringify(property)}</p> */}
                                        {/* Debugging */}
                                        {/* {JSON.stringify(val)} */}
                                    </div>
                                );
                            }

                            // Handle all simple inputs
                            // @ts-ignore
                            const value = props.value[property.identifier];
                            return (
                                <ConfigurationInputFormGroup
                                    card
                                    property={property}
                                    key={property.identifier}
                                >
                                    <ConfigurationInputChild
                                        value={value}
                                        property={property}
                                        onChange={(
                                            updatedValue: OptionValue,
                                        ) => {
                                            props.onChange({
                                                ...props.value,
                                                [property.identifier]: updatedValue,
                                            });
                                        }}
                                    />
                                    {/* Debugging */}
                                    {/* {JSON.stringify(props.value)} */}
                                </ConfigurationInputFormGroup>
                            );
                        },
                    )}
                </ConfigurationGroupVariant>
            </div>
        </div>
    );
};
