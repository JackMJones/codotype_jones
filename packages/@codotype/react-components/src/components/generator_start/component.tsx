import * as React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faPlus } from "@fortawesome/free-solid-svg-icons";
import { GeneratorMeta } from "../types";

// // // //

interface GeneratorStartProps {
    generator: GeneratorMeta;
    externalLink?: string;
    buildLink?: string;
}

/**
 * GeneratorStart
 * Renders the card-header for AttributeEditor and RelationEditor
 * @param props - see `GeneratorStartProps`
 */
export function GeneratorStart(props: GeneratorStartProps) {
    return (
        <div className="row h-100 align-items-center">
            <div className="col-lg-12">
                <div className="row">
                    <div className="col-lg-12 text-center">
                        {/* <img style={{ maxWidth: "4rem !important" }} src={props.label.icon}/> */}
                        <h3 className="display-4">{props.generator.label}</h3>
                        <p className="lead text-muted">
                            powered by{" "}
                            <a target="_blank" href="https://codotype.org">
                                Codotype
                            </a>
                        </p>
                    </div>

                    <div className="col-lg-12 d-flex justify-content-center">
                        {/* <gh-btns-star slug="codotype/codotype" show-count></gh-btns-star> */}
                        {/* GitHub Star Count */}
                        {/* <!-- Place this tag where you want the button to render. --> */}
                        <a
                            className="github-button"
                            href="https://github.com/sponsors/ntkme"
                            data-icon="octicon-heart"
                            aria-label="Sponsor @ntkme on GitHub"
                        >
                            Sponsor
                        </a>

                        {/* <!-- Place this tag where you want the button to render. --> */}
                        <a
                            href="https://github.com/codotype/codotype"
                            data-icon="octicon-star"
                            data-size="large"
                            data-show-count="true"
                            aria-label="Star codotype/codotype on GitHub"
                            className="github-button"
                        >
                            Star
                        </a>
                    </div>
                </div>

                {/* <FontAwesomeIcon icon={faPlus} /> */}

                <div className="row mt-3">
                    <div className="col-lg-12 text-center">
                        <p className="lead">{props.generator.description}</p>
                    </div>
                </div>

                <div className="row justify-content-center">
                    <div className="col-lg-4 col-md-6 col-sm-8 text-center">
                        {props.buildLink && (
                            <button className="btn btn-block btn-primary btn-lg rounded-pill">
                                Let's build something
                                {/* <ChevronAnimation active :svg="svg" /> */}
                            </button>
                        )}

                        {props.externalLink && (
                            <button
                                // v-if="externalLink"
                                // :href="externalLink"
                                className="btn btn-block btn-primary btn-lg rounded-pill"
                            >
                                Let's build something
                                {/* <ChevronAnimation active :svg="svg" /> */}
                            </button>
                        )}
                    </div>
                </div>

                <div className="row row justify-content-center mt-3">
                    <div className="col-lg-12 d-flex justify-content-center align-items-center">
                        {/* <GeneratorTechTag v-for="tag in model.tech_tags" :key="tag" :tag="tag" /> */}
                        Meh
                    </div>

                    <div className="col-lg-12 d-flex justify-content-center align-items-center mt-2">
                        {/* <GeneratorTypeTag v-for="tag in model.type_tags" :key="tag" :tag="tag" /> */}
                        {/* <GeneratorVersionTag :model="model" /> */}
                        {/* <span class='badge badge-info' v-if="model.self_configuring">Self-Configuring</span> */}
                        Content
                    </div>
                </div>
            </div>
        </div>
    );
}

// import GeneratorTypeTag from './GeneratorTypeTag'
// import GeneratorTechTag from './GeneratorTechTag'
// import GeneratorVersionTag from './GeneratorVersionTag'
// import ChevronAnimation from '../../../components/ChevronAnimation'

// export default {
//   props: {
//     model: {
//       required: true
//     },
//     buildLink: {
//       type: String
//     },
//     externalLink: {
//       type: String
//     },
//     svg: {
//       type: Boolean
//     }
//   },
//   components: {
//     GeneratorTypeTag,
//     GeneratorTechTag,
//     GeneratorVersionTag,
//     ChevronAnimation
//   }
// } */}
