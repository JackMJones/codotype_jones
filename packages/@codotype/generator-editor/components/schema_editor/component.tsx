import { SchemaEditorState, schemaEditorReducer, getInitialState, defaultSchema } from "./reducer";
import * as React from "react";
import { Dispatch } from "react";
import { SchemaEditorAction, SchemaEditorActionType } from "./actions";
import { SchemaList } from "./SchemaList";
import { SchemaForm } from "../schema_form";
import { Schema } from "@codotype/types";
import { SchemaListEmptyState } from "./SchemaListEmptyState";

interface SchemaEditorProps {
  schemas: Schema[];
  onChange: (updatedSchemas: Schema[]) => void;
}

export function SchemaEditor(props: SchemaEditorProps) {
  const [state, dispatch]: [SchemaEditorState, Dispatch<SchemaEditorAction>] = React.useReducer(schemaEditorReducer, getInitialState(props.schemas));

  React.useEffect(() => {
    if (state.lastUpdatedTime !== null) {
      props.onChange(state.schemas)
    }
  }, [state.lastUpdatedTime]);

  return (
    <div className="row">
      <div className="col-lg-12">

        <button
          onClick={() => {
            dispatch({
              type: SchemaEditorActionType.EDIT,
              schema: defaultSchema
            })
          }}
        >
          New +
        </button>

        {state.schemas.length === 0 && (
          <SchemaListEmptyState />
        )}

        { state.schemas.length > 0 && (
          <SchemaList
            schemas={state.schemas}
            onEditButtonClick={(schemaToBeEdited: Schema) => {
              dispatch({
                type: SchemaEditorActionType.EDIT,
                schema: schemaToBeEdited
              })
            }}
            onRemoveButtonClick={(schemaToBeRemoved: Schema) => {
              dispatch({
                type: SchemaEditorActionType.REMOVE,
                schema: schemaToBeRemoved
              })
            }}
          />
        )}

        {state.editorModel !== null && (
          <SchemaForm
            schema={state.editorModel}
            onSubmit={(updatedSchema: Schema) => {
              // Creates a new schema
              if (updatedSchema.id === null) {
                dispatch({
                  type: SchemaEditorActionType.CREATE,
                  schema: updatedSchema
                })
                return;
              }

              // Updates an existing schema
              dispatch({
                type: SchemaEditorActionType.UPDATE,
                schema: updatedSchema
              })
            }}
            onCancel={() => {
              dispatch({
                type: SchemaEditorActionType.CLEAR_EDITOR,
              })
            }}
          />
        )}
      </div>
    </div>
  );
}
