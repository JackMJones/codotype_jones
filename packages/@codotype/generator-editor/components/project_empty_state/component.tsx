import * as React from "react";
import { SchemaForm } from "../schema_form";
import { defaultSchema } from "../schema_editor/reducer"

// // // //

/**
 * ProjectEmptyState
 * @param props
 */
export function ProjectEmptyState() {
  return (
    <div className="card card-body">
      <div className='row align-items-center justify-content-center'>
        <div className="col-lg-12">
          <SchemaForm
            schema={defaultSchema}
            onCancel={() => console.log('cancel')}
            onSubmit={() => console.log('submit')}
          />
        </div>
      </div>
      <div className="col-lg-12 text-right">
        <button
          className="btn btn-primary"
        >
          Create Schema
        </button>
      </div>
    </div>
  );
}
