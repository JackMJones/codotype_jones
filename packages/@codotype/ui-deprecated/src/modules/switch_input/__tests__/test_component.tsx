import * as React from "react";
import TestRenderer from "react-test-renderer";
import { SwitchInput } from "../component";

describe("SwitchInput", () => {
  test("render", () => {
    const tree = TestRenderer.create(
      <SwitchInput value={true} onChange={jest.fn()} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
