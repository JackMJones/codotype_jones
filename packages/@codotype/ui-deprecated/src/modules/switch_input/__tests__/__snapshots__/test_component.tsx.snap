// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchInput render 1`] = `
<input
  checked={true}
  onChange={[Function]}
  type="checkbox"
/>
`;
